pallets/permissions/src/lib.rs

```diff
@@ -0,0 +1,281 @@
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+#![cfg_attr(not(feature = "std"), no_std)]
+
+extern crate frame_system;
+
+///! A crate that defines a simple permissions logic for our infrastructure.
+pub use pallet::*;
+
+#[cfg(test)]
+mod mock;
+
+#[cfg(test)]
+mod tests;
+
+#[cfg(feature = "runtime-benchmarks")]
+mod benchmarking;
+
+use frame_support::traits::Contains;
+use frame_support::{dispatch::DispatchResult, pallet_prelude::*};
+use frame_system::pallet_prelude::*;
+
+pub trait Permissions<AccountId> {
+	type Location;
+	type Role;
+	type Error;
+
+	fn has_permission(location: Self::Location, who: AccountId, role: Self::Role) -> bool;
+
+	fn add_permission(
+		editor: AccountId,
+		location: Self::Location,
+		who: AccountId,
+		role: Self::Role,
+	) -> Result<(), Self::Error>;
+
+	fn rm_permission(
+		editor: AccountId,
+		location: Self::Location,
+		who: AccountId,
+		role: Self::Role,
+	) -> Result<(), Self::Error>;
+}
+
+pub trait Properties {
+	type Property;
+
+	fn exists(&self, property: Self::Property) -> bool;
+
+	fn empty(&self) -> bool;
+
+	fn rm(&mut self, property: Self::Property);
+
+	fn add(&mut self, property: Self::Property);
+}
+
+#[frame_support::pallet]
+pub mod pallet {
+	use super::*;
+
+	/// Configure the pallet by specifying the parameters and types on which it depends.
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		/// Because this pallet emits events, it depends on the runtime's definition of an event.
+		type Event: From<Event<Self>> + IsType<<Self as frame_system::Config>::Event>;
+
+		type Location: Member + Parameter;
+
+		type Role: Member + Parameter;
+
+		type Storage: Member + Parameter + Properties<Property = Self::Role> + Default;
+
+		type Editors: Contains<(Self::AccountId, Self::Role)>;
+
+		type AdminOrigin: EnsureOrigin<Self::Origin>;
+	}
+
+	#[pallet::pallet]
+	#[pallet::generate_store(pub(super) trait Store)]
+	pub struct Pallet<T>(_);
+
+	#[pallet::storage]
+	#[pallet::getter(fn permission)]
+	pub type Permission<T: Config> = StorageDoubleMap<
+		_,
+		Blake2_128Concat,
+		T::AccountId,
+		Blake2_128Concat,
+		T::Location,
+		T::Storage,
+	>;
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		RoleAdded(T::AccountId, T::Location, T::Role),
+		RoleRemoved(T::AccountId, T::Location, T::Role),
+		ClearancePurged(T::AccountId, T::Location),
+	}
+
+	// Errors inform users that something went wrong.
+	#[pallet::error]
+	pub enum Error<T> {
+		RoleAlreadyGiven,
+		RoleNotGiven,
+		NoRoles,
+		NoEditor,
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		#[pallet::weight(100)]
+		pub fn add_permission(
+			origin: OriginFor<T>,
+			to: T::AccountId,
+			location: T::Location,
+			role: T::Role,
+		) -> DispatchResult {
+			Self::ensure_admin(origin)?;
```

Comment:
Editors should use the trait interface.

Reply:
I disagree. This will make the extrinsics just used by the Democracy which is not the case 90% of the time. It is mostly, PoolAdmins and MemberlistAdmin who make these changes for a given pool. Letting council or root doing this is just a added benefit

Reply:
Else, Pool Pallet will have to add new extrinsics for pool Admin to add new admins to the pool and restricted tokens will have to implement another extrincisc to add new investors